ParticleObjects/Source/materialSource_class.f90

```diff
@@ -0,0 +1,234 @@
+  use nuclearDataReg_mod,      only : ndReg_getNeutronCE => getNeutronCE, &
+                                      ndReg_getNeutronMG => getNeutronMG
+  use nuclearDatabase_inter,   only : nuclearDatabase
+  use ceNeutronDatabase_inter, only : ceNeutronDatabase
+  use mgNeutronDatabase_inter, only : mgNeutronDatabase
+
+  implicit none
+  private
+
+  !!
+  !! Neutron Source from a given material
+  !!
+  !! Places neutrons uniformly in regions with given material.
+  !! Angular distribution is isotropic.
+  !!
+  !! Can be fed a bounding box to increase sampling efficiency
+  !!
+  !! Private members:
+  !!   isMG   -> is the source multi-group? (default = .false.)
+  !!   bottom -> Bottom corner (x_min, y_min, z_min)
+  !!   top    -> Top corner (x_max, y_max, z_max)
+  !!   E      -> Source site energy [MeV] (default = 1.0E-6)
+  !!   G      -> Source site Group (default = 1)
+  !!
+  !! Interface:
+  !!   source_inter Interface
+  !!
+  !! Sample Dictionary Input:
+  !!   matsource {
+  !!     type materialSource;
+  !!     #data MG; #
+  !!     #E 15.0;  #
+  !!     #G 7;     #
+  !!     #bb (-x -y -z +x +y +z); #
+  !!   }
+  !!
+  type, public,extends(source) :: materialSource
+    private
+    logical(defBool)            :: isMG   = .false.
+    real(defReal), dimension(3) :: bottom = ZERO
+    real(defReal), dimension(3) :: top    = ZERO
+    real(defReal)               :: E      = ZERO
+    integer(shortInt)           :: G      = 0
+    integer(shortInt)           :: matIdx = -1
+  contains
+    procedure :: init
+    procedure :: sampleParticle
+    procedure :: kill
+  end type materialSource
+
+contains
+
+  !!
+  !! Initialise Material Source
+  !!
+  !! See source_inter for details
+  !!
+  subroutine init(self, dict, geom)
+    class(materialSource), intent(inout)     :: self
+    class(dictionary), intent(in)            :: dict
+    class(geometry), pointer, intent(in)     :: geom
+    character(nameLen)                       :: type
+    character(nameLen)                       :: matName 
+    class(nuclearDatabase), pointer          :: nucData
+    class(charMap), pointer                  :: matMap
+    real(defReal), dimension(6)              :: bounds
+    real(defReal), dimension(:), allocatable :: tempArray
+    character(100), parameter :: Here = 'init (materialSource_class.f90)'
+
+    ! Provide geometry info to source
+    self % geom => geom
+
+    ! Select Energy Type
+    call dict % getOrDefault(type, 'data', 'ce')
+    select case(type)
+      case('ce')
+        self % isMG = .false.
+
+      case('mg')
+        self % isMG = .true.
+
+      case default
+        call fatalError(Here, 'Invalid source data type specified: must be ce or mg')
+    end select
+
+    ! Select Required fission group/energy
+    call dict % getOrDefault(self % E, 'E', 1.0E-6_defReal)
+    call dict % getOrDefault(self % G, 'G', 1)
+
+    ! Determine which material to sample source in
+    call dict % get(matName, 'mat')
```

Comment:
You can replace lines 105-113 with a call to materialMenu -> matIdx, that links a name to a material index.